lib/ace/mode/folding/sql.js

```diff
@@ -0,0 +1,52 @@
+/* ***** BEGIN LICENSE BLOCK *****
+ * Distributed under the BSD license:
+ *
+ * Copyright (c) 2010, Ajax.org B.V.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in the
+ *       documentation and/or other materials provided with the distribution.
+ *     * Neither the name of Ajax.org B.V. nor the
+ *       names of its contributors may be used to endorse or promote products
+ *       derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL AJAX.ORG B.V. BE LIABLE FOR ANY
+ * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * ***** END LICENSE BLOCK ***** */
+
+define(function(require, exports, module) {
+"use strict";
+
+var oop = require("../../lib/oop");
+var BaseFoldMode = require("./cstyle").FoldMode;
+
+var FoldMode = exports.FoldMode = function() {};
```

Comment:
Do you think sql fold mode will need to implement some methods different from cstyle? If no, would it be better to use folding/cstyle in sql directly?

Reply:
I assume you mean changing the line in lib/ace/mode/sql.js, from:
`var SqlFoldMode = require("./folding/sql").FoldMode;`
to:
`var SqlFoldMode = require("./folding/cstyle").FoldMode;`

That would be more straightforward now, yes - saves a file.

To make the folder more SQL friendly, we could add support for CASE/END, BEGIN/END or CREATE (or replace function..), but I'm not sure if the latter has a common end marker. Each engine has a slightly different way of creating stored procedures - they mostly start with "CREATE", some end with an "END", others just with a ";".

CASE/END and BEGIN/END should be safe bets in any SQL dialect.

TBH, even cstyle is overkill for what a standard SQL folder needs.
Standard SQL does not have curly/square braces, so folding on round brackets gets us a very useful folder. Even when creating procedures, you usually wrap the body in round brackets, so it would help there as well.

You will find curly braces in SQL + code though, so it might be useful to just let it in there.

Reply:
let's keep the separate file then. Thanks!

Reply:
I had a go at making it work with both brackets and CASE/BEGIN + END, but wasn't succesful.

The "cstyle" folder handles brackets, while the "sqlserver" one can't do brackets, but can handle CASE/BEGIN + END.

Tried merging the two, but it's out of my league atm.